source/common/router/router.cc

```diff
@@ -103,16 +103,13 @@ bool FilterUtility::shouldShadow(const ShadowPolicy& policy, Runtime::Loader& ru
     return false;
   }
 
-  if (policy.defaultValue().numerator() > 0) {
+  if (!policy.runtimeKey().empty()) {
+    // The default is 0/100, which means that the default for a key set but no default value
+    // explicitly set is 0% mirrored.
     return runtime.snapshot().featureEnabled(policy.runtimeKey(), policy.defaultValue(),
                                              stable_random);
   }
 
-  if (!policy.runtimeKey().empty() &&
-      !runtime.snapshot().featureEnabled(policy.runtimeKey(), 0, stable_random, 10000UL)) {
-    return false;
-  }
-
   return true;
```

Comment:
Does this means when the runtime key is empty, then 100% to mirror the request?

A little confused on the default value is required, and runtime key is optional
https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/base.proto#config-core-v3-runtimefractionalpercent

Then I can have a default value, but a empty runtime key. But the default value doesn't work, in the end it is 100%.

Reply:
This is the default behavior today per the documentation. If there is no runtime configured, it is 100% mirrored. 

I don't believe I have changed the default behavior at all, just made it work correctly for lookup via a fractional percent runtime key. Please correct me if this is not correct.

Reply:
I did a test, it seems a behavior change.

I tested with
```
                route:
                  cluster: my_cluster
                  request_mirror_policies:
                    cluster: my_cluster_shadow
                    runtime_fraction:
                      default_value:
                        numerator: 50
                        denominator: HUNDRED
```

In the old code, the default_value is working.

But with this change, it turns to 100% mirror.

The document said if `runtime_fraction` isn't specified, it will be 100% https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#config-route-v3-routeaction-requestmirrorpolicy

But it didn't say that it will be 100% if I specified a default value and a empty runtime key

Reply:
@mattklein123 @snowp ^

Reply:
Hmm ok thanks for the follow up. I will fix this tomorrow.

Reply:
https://github.com/envoyproxy/envoy/pull/20463 thanks again for catching this